### PR TITLE
Fix `SegmentEncoderTransform` to pass inference tests

### LINKED ISSUE #1100: [BUG] Fix `SegmentEncoderTransform` to pass inference tests
### 🐛 Bug Report

According to #1094, `SegmentEncoderTransform` has problems with working on subset of segments and given uncomprehensive error if new segments are met.

### Expected behavior

* `SegmentEncoderTransform` works fine on subset of segments.
* `SegmentEncoderTransform` gives understandable error message if new segments are met.

To make this we should avoid using `self._le.classes_` and use segments present in `df` instead.

### How To Reproduce

Run #1094.

### Environment

_No response_

### Additional context

_No response_

### Checklist

- [X] Bug appears at the latest library version

### PR DESCRIPTION
**Pull request by Mr-Geekman**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Proposed Changes
Look #1100.

## Closing issues
Closes #1100.

### CODE REVIEW

File: etna/transforms/encoders/segment_encoder.py [hunk 1]
```diff
@@ -44,12 +44,31 @@ def transform(self, df: pd.DataFrame) -> pd.DataFrame:
...
+        segments = df.columns.get_level_values("segment").unique().tolist()
+
+        try:
+            new_segments = set(segments) - set(self._le.classes_)
+        except AttributeError:
+            raise ValueError("The transform isn't fitted!")
+
+        if len(new_segments) > 0:
+            raise ValueError(
+                f"This transform can't process segments that weren't present on train data: {new_segments}"
```
Comment (brsnw250) [inline etna/transforms/encoders/segment_encoder.py:64]: Maybe we shouldn't include `new_segments` in error message. If one passes a completely different dataset with a large number of segments, the message will be cluttered.

File: etna/transforms/encoders/segment_encoder.py [hunk 2]
```diff
@@ -44,12 +44,31 @@ def transform(self, df: pd.DataFrame) -> pd.DataFrame:
...
+            )
+
+        encoded_matrix = self._le.transform(segments)
+        encoded_matrix = encoded_matrix.reshape(len(segments), -1).repeat(len(df), axis=1).T
```
Comment (brsnw250) [inline etna/transforms/encoders/segment_encoder.py:68]: If I understood correctly, we can do here: `np.tile(encoded_matrix, (len(df), 1))`

File: etna/transforms/encoders/segment_encoder.py (hunk 1 above)
Comment (Mr-Geekman) [reply etna/transforms/encoders/segment_encoder.py:64]: May be we should try: https://docs.python.org/3/library/reprlib.html

File: etna/transforms/encoders/segment_encoder.py (hunk 2 above)
Comment (Mr-Geekman) [reply etna/transforms/encoders/segment_encoder.py:68]: I will try this. Initially, my intent was to make the smallest change.

**Review by brsnw250 — APPROVED** [review]
👍

**Pull request merged**